### PR TITLE
do some boyscouting in ipam resource

### PR DESCRIPTION
**Pull request by xh3b4sd**
Towards Node Pools.

### CODE REVIEW

File: service/controller/clusterapi/v29/resource/ipam/create.go [hunk 1]
```diff
@@ -85,49 +83,65 @@ func (r *Resource) getReservedNetworks(ctx context.Context) ([]net.IPNet, error)
 	g := &errgroup.Group{}
 
 	g.Go(func() error {
-		r.logger.LogCtx(ctx, "level", "debug", "message", "finding allocated subnets from VPCs")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding allocated subnets from AWSConfig CRs")
 
-		subnets, err := getVPCSubnets(ctx)
+		subnets, err := r.getSubnetsFromAWSConfigs(ctx)
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v29/resource/ipam/create.go]: I re-ordered and re-named the methods.

File: service/controller/clusterapi/v29/resource/ipam/create.go [hunk 2]
```diff
@@ -85,49 +83,65 @@ func (r *Resource) getReservedNetworks(ctx context.Context) ([]net.IPNet, error)
...
 
 		return nil
 	})
 
 	g.Go(func() error {
-		r.logger.LogCtx(ctx, "level", "debug", "message", "finding allocated subnets from Cluster CRs")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding allocated subnets from MachineDeployment CRs")
 
-		subnets, err := getClusterSubnets(r.cmaClient)
+		subnets, err := r.getSubnetsFromMachineDeployments(ctx)
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v29/resource/ipam/create.go]: This one got actually added. We want to also collect subnets from machine deployments.

File: service/controller/clusterapi/v29/resource/ipam/create.go [hunk 3]
```diff
@@ -142,50 +156,39 @@ func (r *Resource) getReservedNetworks(ctx context.Context) ([]net.IPNet, error)
...
-	}
-}
-
-func (r *Resource) splitAndPersistReservedSubnet(ctx context.Context, cr cmav1alpha1.Cluster, subnet net.IPNet, azs []string) error {
-	r.logger.LogCtx(ctx, "level", "debug", "message", "updating CR status to persist network allocation and chosen availability zones")
-
-	var providerStatus g8sv1alpha1.AWSClusterStatus
-
-	err := json.Unmarshal(cr.Status.ProviderStatus.Raw, &providerStatus)
+func (r *Resource) getSubnetsFromAWSConfigs(ctx context.Context) ([]net.IPNet, error) {
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v29/resource/ipam/create.go]: The diff is shite. Functionality did not change. Only the interface changed so it is consistent across all lookup functions.

File: service/controller/clusterapi/v29/resource/ipam/resource.go [hunk 4]
```diff
@@ -5,10 +5,11 @@ import (
 	"reflect"
 
 	"github.com/giantswarm/apiextensions/pkg/clientset/versioned"
-	"github.com/giantswarm/aws-operator/service/network"
 	"github.com/giantswarm/microerror"
 	"github.com/giantswarm/micrologger"
 	"sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset"
+
+	"github.com/giantswarm/aws-operator/service/network"
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v29/resource/ipam/resource.go]: Boyscouting.

File: service/controller/clusterapi/v29/resource/ipam/create.go [hunk 5]
```diff
@@ -85,49 +83,65 @@ func (r *Resource) getReservedNetworks(ctx context.Context) ([]net.IPNet, error)
...
 		reservedSubnets = append(reservedSubnets, subnets...)
 		mutex.Unlock()
 
-		r.logger.LogCtx(ctx, "level", "debug", "message", "found allocated subnets from Cluster CRs")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "found allocated subnets from MachineDeployment CRs")
+
+		return nil
+	})
+
+	g.Go(func() error {
```
Comment (MarcelMue) [inline service/controller/clusterapi/v29/resource/ipam/create.go]: This is an addition, right?

Comment (xh3b4sd) [reply service/controller/clusterapi/v29/resource/ipam/create.go]: No this is the weird diff. We already fetch subnets from the VPCs. What I added is fetching subnets from `MachineDeployment` CRs. Currently the key function is used to get it from the annotations. @tuommaki may wants to check if this is right.

File: service/controller/clusterapi/v29/resource/ipam/create.go [hunk 6]
```diff
@@ -248,3 +251,38 @@ func getVPCSubnets(ctx context.Context) ([]net.IPNet, error) {
 
 	return results, nil
 }
+
+func (r *Resource) persistAllocatedNetwork(cr cmav1alpha1.Cluster, azs []string) func(ctx context.Context, subnet net.IPNet) error {
+	return func(ctx context.Context, subnet net.IPNet) error {
+		return r.splitAndPersistReservedSubnet(ctx, cr, subnet, azs)
```
Comment (tuommaki) [inline service/controller/clusterapi/v29/resource/ipam/create.go]: Would it make sense to just inline that function body here? It's not performing splitting anymore as that will be done by another resource later in the execution.

File: service/controller/clusterapi/v29/resource/ipam/create.go (hunk 3 above)
Comment (tuommaki) [reply service/controller/clusterapi/v29/resource/ipam/create.go]: The diff is indeed total garbage, but luckily split-screen diff makes it somewhat readable :smile:

File: service/controller/clusterapi/v29/resource/ipam/create.go [hunk 7]
```diff
@@ -201,15 +204,15 @@ func getAWSConfigSubnets(g8sClient versioned.Interface) ([]net.IPNet, error) {
 	return results, nil
 }
 
-func getClusterSubnets(cmaClient clientset.Interface) ([]net.IPNet, error) {
-	clusterList, err := cmaClient.Cluster().Clusters(metav1.NamespaceAll).List(metav1.ListOptions{})
+func (r *Resource) getSubnetsFromMachineDeployments(ctx context.Context) ([]net.IPNet, error) {
+	clusterList, err := r.cmaClient.Cluster().MachineDeployments(metav1.NamespaceAll).List(metav1.ListOptions{})
```
Comment (tuommaki) [inline service/controller/clusterapi/v29/resource/ipam/create.go]:
```suggestion
	machineDeploymentList, err := r.cmaClient.Cluster().MachineDeployments(metav1.NamespaceAll).List(metav1.ListOptions{})
```

File: service/controller/clusterapi/v29/resource/ipam/create.go [hunk 8]
```diff
@@ -201,15 +204,15 @@ func getAWSConfigSubnets(g8sClient versioned.Interface) ([]net.IPNet, error) {
...
 	if err != nil {
 		return nil, microerror.Mask(err)
 	}
 
 	var results []net.IPNet
 	for _, c := range clusterList.Items {
```
Comment (tuommaki) [inline service/controller/clusterapi/v29/resource/ipam/create.go]:
```suggestion
	for _, md := range machineDeploymentList.Items {
```

File: service/controller/clusterapi/v29/resource/ipam/create.go [hunk 9]
```diff
@@ -201,15 +204,15 @@ func getAWSConfigSubnets(g8sClient versioned.Interface) ([]net.IPNet, error) {
...
-		cidr := key.StatusClusterNetworkCIDR(c)
+		cidr := key.WorkerSubnet(c)
```
Comment (tuommaki) [inline service/controller/clusterapi/v29/resource/ipam/create.go]:
```suggestion
		cidr := key.WorkerSubnet(md)
```

File: service/controller/clusterapi/v29/resource/ipam/create.go (hunk 6 above)
Comment (xh3b4sd) [reply service/controller/clusterapi/v29/resource/ipam/create.go]: Let's see. I want to talk to you anyway about this thing later.

**Pull request merged**